syntax = "proto3";

package pachyderm.cdr;

option go_package = "github.com/pachyderm/cdr";

message Ref {
    oneof body {
        // Sources
        HTTP http = 1;
        
        // Constraints 
        ContentHash content_hash = 32;
        SizeLimits size_limits = 33;
        
        // 1:1 Transforms
        Cipher cipher = 64;
        Compress compress = 65;
        Slice slice = 66;

        // Many:1 Transforms
        Concat concat = 96;
    }
}

message HTTP {
    string url = 1;
    map<string, string> headers = 2;
}

// Contraints
message ContentHash {
    Ref inner = 1;
    HashAlgo algo = 2;
    bytes hash = 3;
}

// Uses numeric values from this table
// https://github.com/multiformats/multicodec/blob/master/table.csv
enum HashAlgo {
    UNKNOWN_HASH = 0;
    BLAKE2b_256 = 45600;
}

message SizeLimits {
    Ref inner = 1;
    int64 min = 2;
    int64 max = 3;
}

// 1:1 Transforms
message Cipher {
    Ref inner = 1;
    EncAlgo algo = 2;

    bytes key = 3;
    bytes nonce = 4;
}

enum EncAlgo {
    UNKNOWN_CIPHER = 0;
    CHACHA20 = 1;
}

message Compress {
    Ref inner = 1;
    CompressAlgo algo = 2;    
}

enum CompressAlgo{
    UNKNOWN_COMPRESS = 0;
    GZIP = 1;
}

// 1:1 Transforms
message Slice {
    Ref inner = 1;
    uint64 start = 2;
    uint64 end = 3;
}

// Many:1 Transforms
message Concat {
    repeated Ref refs = 1;
}
